Fixed incorrect colour in ErrorBar when Nan value is presented

henryhu123 opened the pull request:
## PR Summary
Fixed incorrect colour in ErrorBar when Nan value is presented
See #13799
Co-author : Dennis Tismenko <Dennis.Tismenko@mail.utoronto.ca>

After analyzing the codebase, the inconsistency issue was traced to the difference between the handling of NaN values in Axes.scatter versus Axes.hlines/Axes.vlines. In the latter case, NaN values were being deleted from their respective np.array, whereas in the former case, NaN values were handled by using a np.ma.array (masked array), where the NaN values were simply masked.

Removed cbook.delete_masked_points() when passing X and Y to Errorbar.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

tacaswell commented:
Overall this seems like the correct fix :+1: 

I am a bit concerned about the images that changed but look identical.  Are there very small changes in them?

For the tests would it be possible to use `check_figures_equal` instead of adding new images? 

See https://matplotlib.org/api/testing_api.html#matplotlib.testing.decorators.check_figures_equal and  https://matplotlib.org/devel/testing.html#writing-an-image-comparison-test

tacaswell commented:
Could you please also simplify / sqaush the git history?

henryhu123 commented:
> Overall this seems like the correct fix 👍
> 
> I am a bit concerned about the images that changed but look identical. Are there very small changes in them?
> 
> For the tests would it be possible to use `check_figures_equal` instead of adding new images?
> 
> See https://matplotlib.org/api/testing_api.html#matplotlib.testing.decorators.check_figures_equal and https://matplotlib.org/devel/testing.html#writing-an-image-comparison-test

henryhu123 commented:
> Overall this seems like the correct fix 👍
> 
> I am a bit concerned about the images that changed but look identical. Are there very small changes in them?
> 
> For the tests would it be possible to use `check_figures_equal` instead of adding new images?
> 
> See https://matplotlib.org/api/testing_api.html#matplotlib.testing.decorators.check_figures_equal and https://matplotlib.org/devel/testing.html#writing-an-image-comparison-test

Sorry accidentally closed the PR. Actually I didn't modify the images for hlines_basics, hlines_with_nan, hlines_masked, vlines_basics, vlines_with_nan, vlines_masked. Not sure why GitHub shows it is modified, I will look into at this issue. Thank you for your feedback.

henryhu123 commented:
> > Overall this seems like the correct fix 👍
> > I am a bit concerned about the images that changed but look identical. Are there very small changes in them?
> > For the tests would it be possible to use `check_figures_equal` instead of adding new images?
> > See https://matplotlib.org/api/testing_api.html#matplotlib.testing.decorators.check_figures_equal and https://matplotlib.org/devel/testing.html#writing-an-image-comparison-test
> 
> Sorry accidentally closed the PR. Actually I didn't modify the images for hlines_basics, hlines_with_nan, hlines_masked, vlines_basics, vlines_with_nan, vlines_masked. Not sure why GitHub shows it is modified, I will look into at this issue. Thank you for your feedback.

Re-uploaded the original test images, test images are not changed now. Can we do the squash merge?

DennisTismenko commented:
@henryhu123 I squashed the git history, so I don't think there's a need for a squash merge.

Included changes to the tests to use `check_figures_equal` instead of image comparison, thank you for the good suggestion.

QuLogic commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -1122,15 +1122,17 @@ def hlines(self, y, xmin, xmax, colors='k', linestyles='solid',
+        # convert verts to MaskedArry.
+        masked_verts = [np.ma.array(
+            [y_xmin, y_xmax])for y_xmin, y_xmax in zip(
+                        np.ma.array([xmin, y]).T, np.ma.array([xmax, y]).T)]
```
The wrapping is really weird here. I also don't understand the comment, as the previous code didn't delete any points either, the `cbook.delete_masked_points` did.

QuLogic commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -1200,15 +1202,17 @@ def vlines(self, x, ymin, ymax, colors='k', linestyles='solid',
+        # convert verts to MaskedArry.
+        masked_verts = [np.ma.array(
+            [xymin, xymax])for xymin, xymax in zip(
+            np.ma.array([x, ymin]).T, np.ma.array([x, ymax]).T)]
```
Same comment here as above.

QuLogic commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -3992,6 +3992,78 @@ def test_hlines():
+    y2 = [2, -6, 3, 8, 2]
+    # Reference image
+    fig_ref, ax2 = plt.subplots()
+    ax2.vlines(y2, 0, x2, colors=colors2, linewidth=5)
```
This function is `hlines`, but you're calling `vlines`. At least the `hlines` vs `vlines` tests could be parametrized, though it would be a bit flaky without #16693.

QuLogic commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -3992,6 +3992,78 @@ def test_hlines():
+    y2 = np.asarray([2, 4, 6])
+    xmin2 = np.asarray([0, 1, -1])
+    xmax2 = np.asarray([13, 14, 15])
+    ax2.vlines(y2, xmin2, xmax2, colors=colors2, linewidth=2)
```
Same here.

DennisTismenko commented:
@QuLogic Thank you for your feedback, I implemented both the changes to the commit message and the code wrapping. We also implemented your suggestions for the tests and fixed the incorrect method calls.

QuLogic commented:
I think you might have misunderstood what parametrization is for, and using `cycler`, while cool, is a bit overkill here. Almost all of your cyclers are 1 element except one that is 2, so their product is 2 long. It seems less work just to write out those 2 differing things.

The point of parametrizing is to remove duplication. Most of that code should still be in the test, and only the changing things need to be parametrized. Certainly the expected data is fixed and there's no need to obfuscate it by putting it in a parameter. I was thinking more that you should parametrize over `vlines`/`hlines` originally, though parametrizing over the two test cases is also a good idea.

henryhu123 commented:
> I think you might have misunderstood what parametrization is for, and using `cycler`, while cool, is a bit overkill here. Almost all of your cyclers are 1 element except one that is 2, so their product is 2 long. It seems less work just to write out those 2 differing things.
> 
> The point of parametrizing is to remove duplication. Most of that code should still be in the test, and only the changing things need to be parametrized. Certainly the expected data is fixed and there's no need to obfuscate it by putting it in a parameter. I was thinking more that you should parametrize over `vlines`/`hlines` originally, though parametrizing over the two test cases is also a good idea.

I have removed `cycler` and moved expected data to test cases. Not sure if I understand your last sentence correctly, I removed duplicated code as much as possible, and created one parametrizing generator function for both hlines/vlines test cases.

timhoffm commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -3992,6 +3992,59 @@ def test_hlines():
+
+@pytest.mark.parametrize('kwargs', generate_lines_with_colors_inputs())
+@check_figures_equal(extensions=["png"])
+def test_vlines_with_colors(fig_test, fig_ref, kwargs):
```
You can greatily simplify the tests:

- The only variable part is the data. All other parameters are identical. You can just hard-code them. No need to build dicts and pass them as `kwargs`.
- ymin/ymax can be scalars.
- You can test `hlines` and `vlines` within one figure: 
![image](https://user-images.githubusercontent.com/2836374/77075043-067bcc00-69f2-11ea-8a74-07fba1b864fd.png)

henryhu123 replied:
- Removed passing kwargs, hard-code the identical values. 
- ymin/ymax are scalars now
- hlines and vlines is in one figure with 2 subplots now.

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -1200,15 +1202,17 @@ def vlines(self, x, ymin, ymax, colors='k', linestyles='solid',
+        ymin = np.ma.resize(ymin, x.shape)
+        ymax = np.ma.resize(ymax, x.shape)
+
+        masked_verts = [np.ma.array([xymin, xymax])
```
Simpler and faster:
~~~
        masked_verts = np.ma.empty((len(x), 2, 2))
        masked_verts[:, 0, 0] = x
        masked_verts[:, 1, 0] = x
        masked_verts[:, 0, 1] = ymin
        masked_verts[:, 1, 1] = ymax
~~~

DennisTismenko replied:
Yes, that is definitely cleaner than what I had before, thank you for the suggestion! Pushed.

henryhu123 commented:
@timhoffm Hi, do you have any idea why Travis CI MacOs build test failed, I have looked through error information and have no idea how to fix it.

QuLogic commented:
This is a known issue #16849.

QuLogic commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -1122,15 +1122,19 @@ def hlines(self, y, xmin, xmax, colors='k', linestyles='solid',
+        masked_verts[:, 0, 1] = y
+        masked_verts[:, 1, 1] = y
+        masked_verts[:, 0, 0] = xmin
+        masked_verts[:, 1, 0] = xmax
```
Can you put these in order? It makes more sense to me this way.
```suggestion
        masked_verts[:, 0, 0] = xmin
        masked_verts[:, 0, 1] = y
        masked_verts[:, 1, 0] = xmax
        masked_verts[:, 1, 1] = y
```

QuLogic commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -1200,15 +1204,19 @@ def vlines(self, x, ymin, ymax, colors='k', linestyles='solid',
+        masked_verts[:, 0, 0] = x
+        masked_verts[:, 1, 0] = x
+        masked_verts[:, 0, 1] = ymin
+        masked_verts[:, 1, 1] = ymax
```
Same here:
```suggestion
        masked_verts[:, 0, 0] = x
        masked_verts[:, 0, 1] = ymin
        masked_verts[:, 1, 0] = x
        masked_verts[:, 1, 1] = ymax
```

DennisTismenko commented:
Pushed the suggested order changes to the creation of `masked_verts`.

timhoffm commented:
Thanks to all who have contributed to this!
